src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,482 @@
+struct GeoPackageBinaryHeader
+  magic::UInt16
+  version::UInt8
+  flags::UInt8
+  srs_id::Int32
+  min_x::Float64
+  max_x::Float64
+  min_y::Float64
+  max_y::Float64
+  min_z::Union{Float64,Nothing}
+  max_z::Union{Float64,Nothing}
+  min_m::Union{Float64,Nothing}
+  max_m::Union{Float64,Nothing}
+end
+
+function parse_gpb_header(blob::Vector{UInt8})
+  io = IOBuffer(blob)
+  
+  magic = read(io, UInt16)
+  magic == 0x5047 || error("Invalid GeoPackage binary header")
+  
+  version = read(io, UInt8)
+  flags = read(io, UInt8)
+  
+  srs_id = ltoh(read(io, Int32))
+  
+  envelope_type = flags >> 1 & 0x07
+  has_envelope = envelope_type > 0
+  
+  if has_envelope && envelope_type >= 1
+    min_x = ltoh(read(io, Float64))
+    max_x = ltoh(read(io, Float64))
+    min_y = ltoh(read(io, Float64))
+    max_y = ltoh(read(io, Float64))
+  else
+    # Default envelope when not present
+    min_x = max_x = min_y = max_y = 0.0
+  end
+  
+  # Initialize z and m coordinates based on envelope type
+  if has_envelope && envelope_type >= 2
+    min_z = ltoh(read(io, Float64))
+    max_z = ltoh(read(io, Float64))
+  else
+    min_z = max_z = nothing
+  end
+  
+  if has_envelope && envelope_type >= 3
+    min_m = ltoh(read(io, Float64))
+    max_m = ltoh(read(io, Float64))
+  else
+    min_m = max_m = nothing
+  end
+  
+  header = GeoPackageBinaryHeader(
+    magic, version, flags, srs_id,
+    min_x, max_x, min_y, max_y,
+    min_z, max_z, min_m, max_m
+  )
+  
+  return header, position(io)
+end
+
+function parse_wkb_geometry(blob::Vector{UInt8}, offset::Int, crs_type)
+  io = IOBuffer(blob[offset+1:end])
+  
+  byte_order = read(io, UInt8)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
+  
+  wkb_type = read_value(read(io, UInt32))
+  
+  base_type = wkb_type & 0x0FFFFFFF
+  
+  # Determine coordinate dimensionality from CRS type
+  # Ignore WKB Z/M flags as they contradict the layer's CRS
+  ndims = CoordRefSystems.ncoords(crs_type)
+  
+  # Create parse_coords closure that reads coordinates based on CRS dimensionality
+  parse_coords = io -> read_coords_from_buffer(io, byte_order, ndims)
+
+  if base_type == WKB_POINT
+    return parse_wkb_point(io, parse_coords, crs_type)
+  elseif base_type == WKB_LINESTRING
+    return parse_wkb_linestring(io, parse_coords, crs_type)
+  elseif base_type == WKB_POLYGON
+    return parse_wkb_polygon(io, parse_coords, crs_type)
+  elseif base_type == WKB_MULTIPOINT
+    return parse_wkb_multipoint(io, byte_order, crs_type)
+  elseif base_type == WKB_MULTILINESTRING
+    return parse_wkb_multilinestring(io, byte_order, crs_type)
+  elseif base_type == WKB_MULTIPOLYGON
+    return parse_wkb_multipolygon(io, byte_order, crs_type)
+  else
+    error("Unsupported WKB geometry type: $base_type")
+  end
+end
+
+# Read coordinates from buffer based on CRS dimensionality
+function read_coords_from_buffer(io::IOBuffer, byte_order::UInt8, ndims::Int)
+  read_value = byte_order == 0x01 ? ltoh : ntoh
```

Comment:
Why compute `read_value` inside all these inner functions? Why not pass it in place of `byte_order`?